### PR TITLE
E2E Utils: add setPreferences and editPost utils

### PR DESCRIPTION
**Pull request by WunderBart**
## What?

- Add `editor.setPreferences()` utility

  `Before`
  ```js
  await page.evaluate( () => {
      window.wp.data
          .dispatch( 'core/preferences' )
          .set( 'core/edit-post', 'welcomeGuide', false );
  
      window.wp.data
          .dispatch( 'core/preferences' )
          .set( 'core/edit-post', 'fullscreenMode', false );
  } );
  ```
  
  `After`
  ```js
  await editor.setPreferences( 'core/edit-post', {
      welcomeGuide: false,
      fullscreenMode: false,
  } );
  ```

- Add `admin.editPost()` utility

  `Before`
  ```js
  await admin.visitAdminPage(
      'post.php',
      `post=${ postId }&action=edit`
  );
  ```
  
  `After`
  ```js
  await admin.editPost( postId );
  ```

- Migrate `admin.createNewPost()` to TS

## Testing Instructions
Tests should pass.

### CODE REVIEW

**Comment by WunderBart** [comment]
@kevin940726 @Mamaduka, I'm looking for early feedback on the new utils: `visitPostEditor` and `setPreferences` which I've created to handle some occurring patterns. I'm unsure about the `setPreferences` one though, as it can be used kind of outside the editor context (e.g. [widgets](https://github.com/WordPress/gutenberg/pull/55099/files#diff-21573fbbc8471a60eb5b4cd98e441a742274b7548999fcad9a20a901a1d4d12bR642)). Do you think it makes sense to have it as the editor utility?

**Review by Mamaduka — COMMENTED** [review]
Thank you, @WunderBart! I think these utils make sense; technically, Widgets are using the block editor.

**Comment by WunderBart** [comment]
@Mamaduka @kevin940726, it's ready for review! 🙌

**Review by Mamaduka — COMMENTED** [review]
@WunderBart, I restarted the performance tests again, but failure might be related to the changes here.

File: packages/e2e-test-utils-playwright/src/admin/visit-site-editor.ts [hunk 1]
```diff
@@ -8,64 +8,42 @@ import { addQueryArgs } from '@wordpress/url';
...
-	// Check if the current page has an editor canvas first.
-	if ( ( await this.page.locator( CANVAS_SELECTOR ).count() ) > 0 ) {
-		// The site editor initially loads with an empty body,
-		// we need to wait for the editor canvas to be rendered.
-		await this.page
-			.frameLocator( CANVAS_SELECTOR )
-			.locator( 'body > *' )
-			.first()
-			.waitFor();
-	}
```
Comment (Mamaduka) [inline packages/e2e-test-utils-playwright/src/admin/visit-site-editor.ts:67]: This code removal doesn't look related to the changes proposed in this PR. As discussed in #54911, there were usually good reasons for introducing similar guardrails.

Maybe we should split this into a separate PR. What do you think?

Comment (WunderBart) [reply packages/e2e-test-utils-playwright/src/admin/visit-site-editor.ts:67]: Ugh, nice catch! I actually forgot to restore this part. I wasn't planning on removing it in this PR, only running a single round without it to see how it goes. Restoring it now.

Comment (WunderBart) [reply packages/e2e-test-utils-playwright/src/admin/visit-site-editor.ts:67]: It's good, though, that now we know only perf tests fail without it - these tests require specific handling, so it might be good to move this safeguard there if it's not necessary here for the _regular_ E2Es. Anyway, it's just a note for myself, I guess - let's leave that for another PR.

File: packages/e2e-test-utils-playwright/src/admin/create-new-post.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * Creates new post.
+ *
+ * @param this
+ * @param options Options to create new post.
+ */
+export async function createNewPost(
+	this: Admin,
+	options: NewPostOptions = {}
+) {
+	const query = addQueryArgs( '', {
```
Comment (kevin940726) [inline packages/e2e-test-utils-playwright/src/admin/create-new-post.ts]: Nit: could we just use `URLSearchParams` here? 😅

File: packages/e2e-test-utils-playwright/src/admin/create-new-post.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+		post_type: options.postType,
+		post_title: options.title,
+		content: options.content,
+		excerpt: options.excerpt,
+	} ).slice( 1 );
+
+	await this.visitAdminPage( 'post-new.php', query );
+
+	await this.editor.setPreferences( 'core/edit-post', {
```
Comment (kevin940726) [inline packages/e2e-test-utils-playwright/src/admin/create-new-post.ts:34]: I'm curious to know if we need to do this every time, seems wasteful to me. Could we just do this in `global-setup` with a network request? I know that the preference API doesn't seem stable in e2e tests though 😅,

File: packages/e2e-test-utils-playwright/src/admin/visit-post-editor.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ * Creates new post.
+ *
+ * @param this
+ * @param options Options to create new post.
+ */
+export async function visitPostEditor(
+	this: Admin,
+	options: PostOptions = {}
+) {
+	if ( typeof options.postId === 'undefined' ) {
```
Comment (kevin940726) [inline packages/e2e-test-utils-playwright/src/admin/visit-post-editor.ts]: Now that we have the opportunity to refactor this, I wonder if we should just combine `createNewPost` and `visitPostEditor`?

```js
// Call with `postId`: visit the existing post.
admin.visitPostEditor( postId: number ): Promise<void>;

// No `postId`: visit `post-new.php` instead.
admin.visitPostEditor(): Promise<void>;
```

WDYT?

File: packages/e2e-test-utils-playwright/src/admin/create-new-post.ts (hunk 3 above)
Comment (Mamaduka) [reply packages/e2e-test-utils-playwright/src/admin/create-new-post.ts:34]: I like the idea. The only time the welcome guide needs to be enabled is when we're actually testing it.

File: packages/e2e-test-utils-playwright/src/admin/visit-post-editor.ts (hunk 4 above)
Comment (WunderBart) [reply packages/e2e-test-utils-playwright/src/admin/visit-post-editor.ts]: Combined in https://github.com/WordPress/gutenberg/pull/55099/commits/59720fe6b5072c37063c43887339ecd2653ca003.

File: packages/e2e-test-utils-playwright/src/admin/create-new-post.ts (hunk 2 above)
Comment (WunderBart) [reply packages/e2e-test-utils-playwright/src/admin/create-new-post.ts]: Updated in https://github.com/WordPress/gutenberg/pull/55099/commits/59720fe6b5072c37063c43887339ecd2653ca003 and https://github.com/WordPress/gutenberg/pull/55099/commits/d2aba51c6b6a3bedfcf5fabc23757a44328d472d

File: packages/e2e-test-utils-playwright/src/admin/visit-site-editor.ts (hunk 1 above)
Comment (WunderBart) [reply packages/e2e-test-utils-playwright/src/admin/visit-site-editor.ts:67]: ~Restored in https://github.com/WordPress/gutenberg/pull/55099/commits/9559d8e2e1981a7d9234a793a9fcb74f97d3ee4a~ See comment below

**Comment by WunderBart** [comment]
> @WunderBart, I restarted the performance tests again, but failure might be related to the changes here.

`setPreferences` was timing out because of the large test post content being loaded. I've increased the default action timeout instead of adding an extra wait only because of the perf tests. See https://github.com/WordPress/gutenberg/pull/55099/commits/36b26c2470ea0ba2ac1c4d8063c6e01227ade6e7.

Also, a little bit of a follow-up cleanup related to this change in https://github.com/WordPress/gutenberg/pull/55441

File: packages/e2e-test-utils-playwright/src/admin/create-new-post.ts (hunk 3 above)
Comment (WunderBart) [reply packages/e2e-test-utils-playwright/src/admin/create-new-post.ts:34]: Is there an existing util for setting prefs via a network request? 🤔

File: packages/e2e-test-utils-playwright/src/admin/visit-site-editor.ts (hunk 1 above)
Comment (WunderBart) [reply packages/e2e-test-utils-playwright/src/admin/visit-site-editor.ts:67]: I had to remove it as the slight timing change caused by using `setPreferences` in `visitSiteEditor` revealed that [the part above](https://github.com/WordPress/gutenberg/pull/55099#pullrequestreview-1673128184) is actually faulty. Before, it checked the canvas iframe(s) **before** they were attached, so that check has always been skipped and moved instantly to wait until the loading spinner is hidden. Now, because `setPreferences` [adds a little delay](https://github.com/WordPress/gutenberg/blob/34f93b329f0c23718817febc7fbfca2b8b47b26b/packages/e2e-test-utils-playwright/src/editor/set-preferences.ts#L23) by waiting until the `window.wp.data` is available, all the editor iframes get attached in the meantime, which [ends up violating](https://github.com/WordPress/gutenberg/actions/runs/6562375488/job/17824208331) the locator strict mode:

```
at /home/runner/work/gutenberg/gutenberg/test/e2e/specs/site-editor/patterns.spec.js:165:3
Error:   1) [chromium] › site-editor/patterns.spec.js:137:2 › Patterns › search and filter patterns ───────

    Retry #1 ───────────────────────────────────────────────────────────────────────────────────────
    Error: locator.waitFor: Error: strict mode violation: locator('iframe[title="Editor canvas"i]').locator('visible=true') resolved to 3 elements:
        1) <iframe tabindex="-1" aria-hidden="true" title="Editor c…></iframe> aka locator('#edit-site-patterns-synced-footer')
        2) <iframe tabindex="-1" aria-hidden="true" title="Editor c…></iframe> aka locator('#edit-site-patterns-unsynced-header')
        3) <iframe tabindex="-1" aria-hidden="true" title="Editor c…></iframe> aka locator('#edit-site-patterns-unsynced-footer')
```

ℹ️ The other iframes are previews of the patterns created in that test.

**Comment by WunderBart** [comment]
@Mamaduka @kevin940726 it's ready for another go! 🙏

**Comment by WunderBart** [comment]
@Mamaduka @kevin940726 🏓 🙏

**Comment by Mamaduka** [comment]
Sorry, I was AFK last week and missed the notification. I am adding this to my to-do list and will review it later today.

File: packages/e2e-test-utils-playwright/src/admin/visit-post-editor.ts (hunk 4 above)
Comment (Mamaduka) [reply packages/e2e-test-utils-playwright/src/admin/visit-post-editor.ts]: A few notes on this:

1. The https://github.com/WordPress/gutenberg/commit/59720fe6b5072c37063c43887339ecd2653ca003 is a breaking change. We shouldn't ship that as `@wordpress` packages try to follow the same rules as WP regarding backward compatibility.
2. Personally, I prefer a more explicit `createNewPost` action.

@WunderBart, I guess we can make `createNewPost` a wrapper around `visitPostEditor`. We just need to make sure that the argument signature is the same.

Comment (WunderBart) [reply packages/e2e-test-utils-playwright/src/admin/visit-post-editor.ts]: The wrapper idea makes sense to me. cc @kevin940726 if he has any objections here!

**Comment by kevin940726** [comment]
Sorry I'm also AFK until later next week 😅. Feel free to merge without me though! 🙏

File: packages/e2e-test-utils-playwright/src/admin/visit-post-editor.ts (hunk 4 above)
Comment (WunderBart) [reply packages/e2e-test-utils-playwright/src/admin/visit-post-editor.ts]: > Personally, I prefer a more explicit createNewPost action.

I'm actually starting to lean in this direction as well. The main issue with merging createNewPost and visitPostEditor are the props (and their typing) that can be confusing (e.g. if postId is passed, all the props for creating a new post are ignored). Personally, at this point I also feel like a more explicit approach would be better - keep the `createNewPost` and implement `editPost` instead of `visitPostEditor`. For the Site Editor, it makes sense to be reachable by `visitSiteEditor` because there's only one URL we can visit it by: /site-editor.php. 

Sorry for going back and forth with it, but I just want to avoid overcomplicating stuff (which I might already be doing 😅).  Please let me know if the above would make sense to you!

Comment (Mamaduka) [reply packages/e2e-test-utils-playwright/src/admin/visit-post-editor.ts]: > keep the `createNewPost` and implement `editPost` instead of `visitPostEditor`.

100% - I think this fits the WP glossary around posts much better.

> Sorry for going back and forth with it, but I just want to avoid overcomplicating stuff

No worries. It's easier to anticipate what would work best when we try. Thanks for taking the time and experimenting with different approaches.

**Review by Mamaduka — APPROVED** [review]
Thanks for the follow-ups, @WunderBart. The changes here look good to me 🚢

File: packages/e2e-test-utils-playwright/src/admin/visit-post-editor.ts (hunk 4 above)
Comment (kevin940726) [reply packages/e2e-test-utils-playwright/src/admin/visit-post-editor.ts]: I appreciate the discussions and I agree with the final solution we settled on too! Thank you all 🧡
